lib/srv/git/github.go

```diff
@@ -30,35 +34,102 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/protobuf/types/known/durationpb"
 
+	"github.com/gravitational/teleport"
 	integrationv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/integration/v1"
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/sshutils"
 )
 
-// knownGithubDotComFingerprints contains a list of known GitHub fingerprints.
-//
-// https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints
-//
-// TODO(greedy52) these fingerprints can change (e.g. GitHub changed its RSA
-// key in 2023 because of an incident). Instead of hard-coding the values, we
-// should try to periodically (e.g. once per day) poll them from the API.
-var knownGithubDotComFingerprints = []string{
-	"SHA256:uNiVztksCsDhcc0u9e8BujQXVUpKZIDTMczCvj3tD2s",
-	"SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM",
-	"SHA256:+DiY3wvvV6TuJJhbpZisF/zLDA0zPMSvHdkr4UvCOqU",
+// githubServerKeyManager downloads SSH keys from the GitHub meta API and does a
+// lazy refresh every hour. The keys are used to verify GitHub server when
+// forwarding Git commands to it.
+type githubServerKeyManager struct {
+	mu        sync.Mutex
+	keys      []string
+	lastCheck time.Time
+	etag      string
+
+	clock       clockwork.Clock
+	apiEndpoint string
+	client      *http.Client
+}
+
+func newGitHubServeKeyManager() *githubServerKeyManager {
+	return &githubServerKeyManager{
+		clock:       clockwork.NewRealClock(),
+		apiEndpoint: "https://api.github.com/meta",
+		client: &http.Client{
+			Timeout: defaults.HTTPRequestTimeout,
+		},
+	}
```

Comment:
We start with an empty list of keys that will only be refreshed on first request. This increases the latency and has unhappy consequences on error. Given how often GitHub experiences hiccups, I'd prefer to run `refreshLocked` in a loop independently of the `check` method, with variable schedule based on current status. 

For example:
1. Empty list of known keys? Retry every minute until we get the list.
2. Non-empty list of known keys, but last attempt at refresh failed? Retry after 15 minutes.
3. Non-empty list of known keys and last attempt worked? Repeat after 12h.

Reply:
I think the requests are cheap enough it shouldn't matter if we fire a single one every few hours. On the other hand if the API is unavailable (e.g. due to firewall misconfiguration) it would be better to know that sooner rather than later.

Perhaps we can leave it as it is for now, but simply fire a single `go refreshLocked()` on startup?

Reply:
decided to make a proper git.KeyCertManger that Proxy can pass along the way. will update the PR soon